postgres: Move lock out of ensureVersionTable, for consistency with other SQL operations

lukaspj opened the pull request:
Fixes #164 .

`ensureVersionTable` seems to be the only SQL operation in the postgres driver which handles locking of database itself, meaning it seems to break the convention of having the caller responsible for locking and not the callee.

This removes the locking logic from `ensureVersionTable` and makes sure to properly lock the database before calling `ensureVersionTable` from the `WithInstance` method.

dhui commented on database/postgres/postgres.go:
```diff
@@ -98,10 +98,23 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 		return nil, err
 	}
 
+	if e := px.Unlock(); e != nil {
```
Should be deferred immediately after `px.Lock()`

lukaspj commented:
@dhui looking at the MySQL implementation as reference, it simply doesn't lock the database when calling ensureVersionTable from either location.

I think the behaviour should be similar in the two, should I simply remove the locking logic around ensureVersionTable? Thinking of it, it doesn't seem like it'd be necessary to lock it for that operation.

dhui commented:
> I think the behaviour should be similar in the two, should I simply remove the locking logic around `ensureVersionTable` ? Thinking of it, it doesn't seem like it'd be necessary to lock it for that operation.

`ensureVersionTable()` should be atomic. A [recent fix](https://github.com/golang-migrate/migrate/pull/150) was made to fix an [issue with multiple instances of migrate running](https://github.com/golang-migrate/migrate/issues/55), which incidentally created the [issue you're trying to fix](https://github.com/golang-migrate/migrate/issues/164).

> Edit: otherwise, if we want to remove `ensureVersionTable` from `Drop`, I think we should do so in other database implementations as well for consistency.

This work would be greatly appreciated! MySQL isn't the only driver that uses `ensureVersionTable()`:
```
$ git grep -l ensureVersionTable
database/cassandra/cassandra.go
database/clickhouse/clickhouse.go
database/cockroachdb/cockroachdb.go
database/mysql/mysql.go
database/mysql/mysql_test.go
database/postgres/postgres.go
database/ql/ql.go
database/redshift/redshift.go
database/spanner/spanner.go
database/sqlite3/sqlite3.go
$
```

So for consistency, we should lock in `ensureVersionTable()` **and** remove the call to `ensureVersionTable()` from `Drop()`

lukaspj commented:
Alright, I will update the PR tomorrow. 
Another suggestion, should it be renamed to `lockAndEnsureVersionTable` to make the break from the *caller locks* convention explicit?
Might avoid future confusion

dhui commented:
> Another suggestion, should it be renamed to `lockAndEnsureVersionTable` to make the break from the _caller locks_ convention explicit?
Might avoid future confusion

I'd leave the name as is since it's sufficiently descriptive. After all, adding every bit of functionality to the function's name would result in really long function names.
The docs around db driver lock usage are on the the `Driver` interface in `database/driver.go` which *should* be enough to prevent future confusion. But if you have suggestions for improving the docs, I'm all ears.

lukaspj commented:
@dhui I added a test-case, but only for Postgresql, as I don't know enough about most of the other databases.
See: https://github.com/golang-migrate/migrate/pull/173/files#diff-5cc52c39a5031732bc6d3821ffc558e6

Also, please be aware of this change:
https://github.com/golang-migrate/migrate/pull/173/files#diff-9870930c9960513ad7d4eae22e4692d5

The change in behaviour in Drop (namely that it leaves an empty database), breaks some logic in these test-cases (and possibly in user-code as well).

Finally, it's worth mentioning that the order of ensureLock / ensureVersion has been swapped in CockroachDB, in order to facilitate locking in ensureVersion

lukaspj commented:
Btw, it looks like coverage is only measured for MongoDB and not all databases:
https://coveralls.io/builds/21705329

dhui commented on database/postgres/postgres_test.go:
```diff
@@ -307,6 +307,49 @@ func TestPostgres_Lock(t *testing.T) {
 	})
 }
 
+func TestDrop(t *testing.T) {
```
Rename to `TestMigrateDrop` and add a comment that mentions preventing a regression for https://github.com/golang-migrate/migrate/issues/164

It may be worth moving this test to `Test()` in `database/testing/testing.go` to test all DB driver implementations.

lukaspj replied:
Should I leave this test as-is? A new `TestMigrateDrop()` has been added to `database/testing/migrate_testing`, however this test makes sure that the Postgresql driver specifically is able to `Drop()` both version table and tables created through SQL, outside of migrations.

dhui replied:
I don't think this test is necessary anymore since all cases are covered by `dk.TestDrop()` and `dk.TestMigrateDrop()`

lukaspj replied:
Alright then.

dhui commented on database/cockroachdb/cockroachdb.go:
```diff
@@ -85,11 +86,12 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 	}
 
-	if err := px.ensureVersionTable(); err != nil {
+	// ensureVersionTable is a locking operation, so we need to ensureLockTable before we ensureVersionTable.
```
Thanks for finding and fixing this issue!
The comment is useful.

It's unfortunate that lock table creation will have issues when using multiple instances of migrate and cockroachdb and there's nothing we can do about it....

lukaspj replied:
Yeah, I guess it would be possible to lock it on the Driver-level, but that only helps for that specific Driver instance, doesn't help if there are multiple Drivers running on the same database. :(

lukaspj commented on database/postgres/postgres_test.go:
```diff
@@ -307,6 +307,49 @@ func TestPostgres_Lock(t *testing.T) {
+		}
+
+		// drop the table
+		if err := d.Drop(); err != nil {
```
Hmm... The migrate is a parent-package which the database package doesn't know, so it wouldn't make sense to use migrate's `Drop()` here. Then the test should be moved to the `migrate_test.go` or another test file in the root package but that doesn't test the specific database implementations.

I guess we would need a suite that does integration-tests on migrate and all the different database implementation which would be a rather significant addition.

I could add such a suite, but it would only cover the `Drop()` method, as I don't have the time to write migrations for each database.

dhui replied:
Yeah, I wanna avoid scope creep, but also want to prevent a regression. We don't need to test the other methods. e.g. only testing `Drop()` is sufficient. The test in it's current form doesn't cover any different cases than the existing `TestDrop()` in `database/testing/testing.go`.

The quickest way (albeit less clean way) to add the proper regression test would be to import `github.com/golang-migrate/migrate/v4` in `database/testing/testing.go`, create a `migrate.Migrate` instance using `NewWithDatabaseInstance()` in `Test()`, and test `Drop()` using that instance.

I'm all for adding a db driver <-> migrate integration test suite, but I'm not sure about the approach you have for this. The scope for such an integration test suite may also be quite large, so for now, let's add the regression test using the quick and dirty way mentioned above. Then we can add a db driver integration test suite and clean the code in a separate PR.

Thanks again for your work!

lukaspj commented:
I went a bit rogue and made some changes.
First of all I cleaned up the git commit history a bit with a force-push.

I found that a change to the Driver interface was warranted. With the change to the functionality in `Drop()` there was no way to re-establish the version table after a `Drop()`. I found this to be quite tedious, at it meant you had to store a copy of the database URL for re-connecting if your code would do a `Drop()`.
To this end I added the `Initialize()` method to the Driver interface:
https://github.com/golang-migrate/migrate/pull/173/commits/f111488a188eb9da566e8774e0503c1165faef51#diff-ea51b70c3e62fa1445d70670eceaafe6

Logic for choosing the default MigrationsTable, was moved to this `Initialize()` method, to ensure that a call to `Initialize()` always had a sensible MigrationsTable.

For the regression test, found here:
https://github.com/golang-migrate/migrate/pull/173/commits/f111488a188eb9da566e8774e0503c1165faef51#diff-8571f16a46121ca4b6bd807b05b6ed75

I wanted to keep it completely separate from the existing tests, so I had to duplicate some code in order to call the `TestMigrate` method.
Example:
https://github.com/golang-migrate/migrate/pull/173/commits/f111488a188eb9da566e8774e0503c1165faef51#diff-30f12fe639ae221117c9c518485c1930

I just duplicated the whole test-block when `dktesting.ParallelTest` was used. When ParallelTest wasn't used, I did a `Drop()` and `Initialize()` before invoking `TestMigrate()`

Example:
https://github.com/golang-migrate/migrate/pull/173/commits/f111488a188eb9da566e8774e0503c1165faef51#diff-da587f4c74d6e6dca490ec5a1807c0f5R43

@dhui what do you think of these changes? Does the change to the Driver interface make sense to you?

dhui commented:
> I went a bit rogue and made some changes.
First of all I cleaned up the git commit history a bit with a force-push.

That's fine since it's already done. In the future, I'd prefer that we squash commits after PR approval but before merging to cleanup the history to make reviewing easier. However your comments summarizing your changes have been quite useful!

> I found that a change to the Driver interface was warranted. With the change to the functionality in `Drop()` there was no way to re-establish the version table after a `Drop()`. I found this to be quite tedious, at it meant you had to store a copy of the database URL for re-connecting if your code would do a `Drop()`.
To this end I added the `Initialize()` method to the Driver interface:

I'd avoid making any changes to the `database.Driver` interface. As you mentioned, you could achieve the same goal in testing by re-connecting. Although such a change is more tedious, it's less intrusive. Any changes to the `database.Driver` interface should be discussed [here](https://github.com/golang-migrate/migrate/issues/14) first.
I don't think we should add an `Initialize()` method to `database.Driver` since it makes driver instantiation more clumsy and more error prone. e.g. it's better keep the interface exposed to users simple

dhui commented on database/testing/migrate_testing.go:
```diff
@@ -0,0 +1,38 @@
+
+// TestMigrate runs integration-tests between the Migrate layer and database implementations.
+//
+func TestMigrate(t *testing.T, d database.Driver, migration []byte) {
```
This test should take a `*migrate.Migrate` instead of a `database.Driver`

dhui commented on database/driver.go:
```diff
@@ -74,6 +74,8 @@ type Driver interface {
 	Version() (version int, dirty bool, err error)
 
 	// Drop deletes everything in the database.
+	// Note that this is a breaking action, a new call to Open() is necessary to
```
Thanks for the comment!

lukaspj commented:
Sorry for abusing your test-server btw and testing-by-pushing, but I can't run Docker locally. I don't have enough RAM in this PC.

lukaspj commented:
Alright, in order to create a `TestMigrate` I changed the tests so that they now read the migrations in the `examples` folder for that test run. 

I aligned the folder structure in the database implementations, such that they all have the migration examples in `examples/migrations/`.

dhui commented on database/mongodb/mongodb.go:
```diff
@@ -77,9 +78,6 @@ func (m *Mongo) Open(dsn string) (database.Driver, error) {
 		return nil, err
 	}
 	migrationsCollection := purl.Query().Get("x-migrations-collection")
-	if len(migrationsCollection) == 0 {
```
default config value fallback removed

lukaspj replied:
Actually I did that on purpose, forgot to mention it I guess. 'Twas a long day.

The reason for this change is that the fallback was set both in `WithInstance` and in `Open`.
From the code, it would seem that it only needs to be set in `WithInstance` as no step in `Open`, before the `WithInstance` test, requires the field.
It could also be moved out of `WithInstance` and into `Open` but I believe it makes sense to have it in `WithInstance`.

dhui commented on database/mysql/mysql.go:
```diff
@@ -75,15 +76,15 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 
 	config.DatabaseName = databaseName.String
 
-	if len(config.MigrationsTable) == 0 {
```
default config value fallback removed

Use original ordering for a cleaner diff. Also, config values should be "set" before any instances or connections are made.

dhui commented on database/mysql/mysql.go:
```diff
@@ -127,9 +128,6 @@ func (m *Mysql) Open(url string) (database.Driver, error) {
 	purl.RawQuery = q.Encode()
 
 	migrationsTable := purl.Query().Get("x-migrations-table")
-	if len(migrationsTable) == 0 {
```
default config value fallback removed

dhui commented on database/postgres/postgres.go:
```diff
@@ -117,14 +122,12 @@ func (p *Postgres) Open(url string) (database.Driver, error) {
 	}
 
 	migrationsTable := purl.Query().Get("x-migrations-table")
-	if len(migrationsTable) == 0 {
```
default config value fallback removed

dhui commented on database/redshift/redshift.go:
```diff
@@ -64,16 +65,16 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 
 	config.DatabaseName = databaseName
 
-	if len(config.MigrationsTable) == 0 {
```
default config value fallback removed

Use original ordering for a cleaner diff. Also, config values should be "set" before any instances or connections are made.

dhui commented on database/redshift/redshift.go:
```diff
@@ -100,9 +101,6 @@ func (p *Redshift) Open(url string) (database.Driver, error) {
 	}
 
 	migrationsTable := purl.Query().Get("x-migrations-table")
-	if len(migrationsTable) == 0 {
```
default config value fallback removed

dhui commented on database/spanner/spanner.go:
```diff
@@ -108,9 +109,6 @@ func (s *Spanner) Open(url string) (database.Driver, error) {
 	}
 
 	migrationsTable := purl.Query().Get("x-migrations-table")
-	if len(migrationsTable) == 0 {
```
default config value fallback removed

dhui commented on database/postgres/postgres.go:
```diff
@@ -92,13 +92,18 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 	}
 
-	if err := px.ensureVersionTable(); err != nil {
+	err = px.ensureVersionTable()
```
Why break this up?

dhui commented on database/postgres/postgres.go:
```diff
@@ -92,13 +92,18 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 		return nil, err
 	}
 
+	if len(config.MigrationsTable) == 0 {
```
This is unnecessary as it's already done above (line 85)

dhui commented on database/ql/ql_test.go:
```diff
@@ -40,23 +40,38 @@ func Test(t *testing.T) {
+	defer func() {
+		os.RemoveAll(dir)
+	}()
+	fmt.Printf("DB path : %s\n", filepath.Join(dir, "ql.db"))
```
Use `t.Logf()` instead

lukaspj replied:
Right, this was a copy-pasta, I will correct it in `Test()` as well.

dhui commented on database/spanner/spanner_test.go:
```diff
@@ -26,3 +28,26 @@ func Test(t *testing.T) {
+	}
+
+	s := &Spanner{}
+	addr := fmt.Sprintf("spanner://%v", db)
```
Use simple string concatenation or `%s` as the formatter

dhui commented:
Thanks for the PR!
